qualtran/_infra/composite_bloq_test.py

```diff
@@ -556,6 +556,25 @@ def test_t_complexity():
     assert TestParallelCombo().t_complexity().t == 3 * 100
 
 
+def test_add_and_partition():
+    from qualtran import Controlled, CtrlSpec
+    from qualtran.bloqs.basic_gates import Swap
+
+    bb = BloqBuilder()
+    bloq = Controlled(Swap(3), CtrlSpec(qdtypes=QUInt(4), cvs=0b0110))
+    a = bb.add_register_from_dtype('a', QAny(7))
+    b = bb.add_register_from_dtype('b', QAny(3))
+    assert a is not None
+    assert b is not None
+    ctrl_r, x_r, y_r = bloq.signature.lefts()
+    a, b = bb.add_and_partition(
+        bloq, [(Register('a', QAny(7)), [y_r, ctrl_r]), (Register('b', QAny(3)), [x_r])], a=a, b=b
```

Comment:
Maybe this would cause problems, but what do you think about an API where we can use the names of the factorized registers to set up the auto-partition. 

```python
bb.add_and_partition(CSwap(), {Register('a', QAny(7)): ['y', 'ctrl']. Register('b', QAny(3): ['x']}, a=a, b=b)
```

Reply:
I think this works. I'll change it.